core/material.js

```diff
@@ -1,26 +1,45 @@
 import { create_program_object, create_shader_object, gl } from './context';
 import { Transform, Render } from '../components';
 
+import { vertex } from './shaders';
+import { fragment } from './shaders';
+
 export class Material {
-  constructor() {
+  constructor(options) {
     this.uniforms = {};
     this.attribs = {};
+    this.features = [];
+    this.features_from_components(Array.from(options.components) || []);
+  }
+
+  features_from_components(components) {
+    this.features = components
+      .reduce((memo, component) => {
+        if (component.features.length > 0) {
+          return memo.concat(component.features);
+        } else {
+          return memo;
+        }
+      }, [])
```

Comment:
You can safely concat an empty array and this code then becomes a one-liner:

```javascript
this.features = components.reduce((acc, cur) => acc.concat(cur.features), []);
```

Or, you could first build an array of features per component and pass them all to a single `concat`:

```javascript
this.features = [].concat(components.map(comp => comp.features));
```

Reply:
The line above should read:

```javascript
this.features = [].concat(...components.map(comp => comp.features));
```